File: 139 (Word Break)/java/Solution.java
class Solution {

    boolean solve(int idx, String s, Set<String> wordset, HashMap<Integer,Boolean> dp){
        if (idx==s.length()) return true;
        if (dp.containsKey(idx)) return dp.get(idx);

        for(int i = idx+1;i<=s.length();i++){
            if (wordset.contains(s.substring(idx,i)) && solve(i,s,wordset,dp)){
                dp.put(idx,true);
                return true;
            }
        }
        dp.put(idx,false);
        return false;
    }
    public boolean wordBreak(String s, List<String> wordDict) {
        Set<String> wordset = new HashSet<>(wordDict);
        HashMap<Integer,Boolean> dp = new HashMap<>();

        return solve(0,s,wordset,dp);
    }
}

File: 139 (Word Break)/javascript/Solution.js
/**
 * @param {string} s
 * @param {string[]} wordDict
 * @return {boolean}
 */
var wordBreak = function(s, wordDict) {
  const wordset = new Set(wordDict);
  const dp = new Map();

  const solve = (idx) => {
      if (idx==s.length) return true;
      if (dp.has(idx)) return dp.get(idx);

      for(let i = idx+1;i<=s.length;i++){
          if (wordset.has(s.substring(idx,i)) && solve(i)){
              dp.set(idx,true);
              return true;
          }
      }
      dp.set(idx,false);
      return false;
  }
  return solve(0);
};